### ISSUE TITLE
Strange segfault on --release

### ISSUE THREAD

**Issue by Exilor** [manastech/crystal#574]
I haven't been able to reproduce it with a smaller example. The code is [here](https://gist.github.com/Exilor/5e848e490e0bf3eb478a).
The segfault seems to appear when [#to_s](https://gist.github.com/Exilor/5e848e490e0bf3eb478a#file-big_float-cr-L204) (which I admit is very ugly) is called under some conditions I can't for the life of me figure out. For example:

```crystal
p 5.to_big_f
p 5.to_big_f
p 5.to_big_f.to_s
# segfault
```
```crystal
p 5.to_big_f
p 5.to_big_f.to_s
# 5
# "5"
# doesn't segfault
```
```crystal
p 5.to_big_f.to_s.to_big_f.to_s
# segfault
```
```crystal
n = 0.0
while n < 10
  puts n.to_big_f
  n += 0.123
end
# doesn't segfault
```

The segfault doesn't happen if I remove the call to ```#chr``` from line [218](https://gist.github.com/Exilor/5e848e490e0bf3eb478a#file-big_float-cr-L218) or compile without --release. Other than that, ```#to_s``` does exactly what I want it to do.

Basically I don't know if my code is at fault, if I found a bug, or both.

**Comment by ysbaddaden** [comment]
Here is a GDB backtrace for the first example (3 times `p 5.to_big_f`):

```
Program received signal SIGSEGV, Segmentation fault.
fputc (c=53, fp=0x7fff00000000) at fputc.c:37
37      fputc.c: Aucun fichier ou dossier de ce type.
(gdb) bt
#0  fputc (c=53, fp=0x7fff00000000) at fputc.c:37
#1  0x0000000000403d9c in *CFileIO@IO#<<<CFileIO, Char>:CFileIO ()
#2  0x00000000004025f8 in main ()
```

`5` isn't even printed on screen once, and there ain't indeed no crash in non release mode. I guess the optimizer is optimizing too much.

**Comment by Exilor** [comment]
@ysbaddaden how did you get that info?

**Comment by jhass** [comment]
Confirmed here (LLVM 3.6). Valgrind reports:
```
==13676== Process terminating with default action of signal 11 (SIGSEGV)
==13676==  Access not within mapped region at address 0x0
==13676==    at 0x60A8C71: fputc (in /usr/lib/libc-2.21.so)
==13676==    by 0x10C17B: *CFileIO@IO#<<<CFileIO, Char>:CFileIO (in /home/jhass/projects/crystal/tests/big_f/big_float)
==13676==    by 0x10C0F9: *p<BigFloat>:BigFloat (in /home/jhass/projects/crystal/tests/big_f/big_float)
==13676==    by 0x10A7F9: main (in /home/jhass/projects/crystal/tests/big_f/big_float)
```

**Comment by ysbaddaden** [comment]
Run the binary through gdb:

```
$ crystal build --release test_bf.cr
$ gdb test_bf
> run
...
> bt
```

**Comment by asterite** [comment]
Well, after a good hour of debugging and trying to reduce it, I came to the conclusion that this:

```crystal
fun mpf_get_str = __gmpf_get_str(str : UInt8*, expptr : Int32*, base : Int32, n_digits : Int32, op : MPF*) : UInt8*
```

must actually be this:

```crysta
fun mpf_get_str = __gmpf_get_str(str : UInt8*, expptr : Int64*, base : Int32, n_digits : Int32, op : MPF*) : UInt8*
```

Note the type of expptr. In the docs that's mp_exp_t, which actually depends on some _CRAY flag. I have no idea what's that, or how to prevent this kind of errors. Having a binding generator will probably generate them with types for the current machine... right?

I guess this will always be trial and error.

**Comment by asterite** [comment]
The conclusion is: it might be a compiler bug, but when writing a C binding, always check the types in the bindings. Because `expptr` is an "out" parameter, the function writes to it and it was probably overwriting some data on the stack or something like that (just a guess). In fact in a small test I had the object_id of an IO instance was changing, which probably means the above is sorta true.

**Comment by Exilor** [comment]
@ysbaddaden Thanks!
@asterite Sorry for making you waste an hour on this. I was so sure that I had the types of mpf_get_str right.

**Comment by asterite** [comment]
Not at all! Now I know that I need to check for the type of `out` parameters besides struct definitions ;-)
